File: 220.immutability/src/main/scala/scalax/Factorial.scala
package scalax

import scala.annotation.tailrec

object Factorial extends App {
  assert(args.size == 1, "Usage: Factorial <n>")
  val n = args(0).toInt
  assert(n >= 0, "<n> needs to be >= 0")

  println(fac(n))
  println(facRec(n))
  println(facTailRec(n))

  private def fac(n: Int): Int = {
    var f = 1
    for (i <- 1 to n) {
      f = f * i
    }
    f
  }

  private def facRec(n: Int): Int = {
    if (n == 0) 1
    else n * facRec(n - 1)
  }

  private def facTailRec(n: Int): Int = {
    @tailrec
    def fac(n: Int, acc: Int): Int = {
    if (n == 0) acc
    else fac(n - 1, n * acc)
    }
    fac(n, 1)
  }
}

File: 220.immutability/src/main/java/Factorial.java
package javax;

public class Factorial {
    public static void main(String[] args) {
        assert(args.length == 1);
        final int n = Integer.parseInt(args[0]);
        assert(n >= 0);

        System.out.println(fac(n));
    }

    private static int fac(final int n) {
        int f = 1;
        for(int i = 1; i <= n; i++) {
            f = f * i;
        }
        return f;
    }
}
